Documentation rewrite - Get started page (#3059)
Kotlin/kotlinx.serialization: docs-website/topics/serialization-get-started.md

[//]: # (title: Get started with Kotlin serialization)

[Serialization](serialization.md) converts objects into a format that can be stored or transmitted and later reconstructed.

This tutorial shows you how to add the necessary plugins and dependencies for Kotlin serialization and how to serialize and deserialize objects in JSON format.

## Add plugins and dependencies for Kotlin serialization

To include the `kotlinx.serialization` library in your project, add the corresponding plugin and dependency configuration based on your build tool.

> > To set up the Kotlin compiler plugin for Bazel, see the example provided in the [rules_kotlin repository](https://github.com/bazelbuild/rules_kotlin/tree/master/examples/plugin/src/serialization).

Reviewer: I think this note should be moved after Gradle/Maven sections

Author: hmm 🤔 my issue with that is that way it might seem like it's related to Maven. Since it's just a note without specific instructions I left it in the generic part of the doc.

Reviewer 2: If there is a lot of stuff on different build systems it even makes sense to have a separate page for it (linked here). It would also help those who want to find instructions for setting it up in different build systems as they often would not expect it in a generic getting started page.

Author: Get started page should be enough — Let's merge Gradle + Maven into Tabs (very compact takes up not too much space)

                       <id>compile</id>
                       <phase>compile</phase>
                       <goals>
                           <goal>compile</goal>
                       </goals>
                   </execution>
               </executions>
               <configuration>
                   <compilerPlugins>
                       <plugin>kotlinx-serialization</plugin>
                   </compilerPlugins>
               </configuration>
               <dependencies>
                   <dependency>
                       <groupId>org.jetbrains.kotlin</groupId>
                       <artifactId>kotlin-maven-serialization</artifactId>
                       <version>${kotlin.version}</version>
                   </dependency>
               </dependencies>
           </plugin>
       </plugins>
   </build>
   ```

3. Add the JSON serialization library dependency to the `<dependencies>` section:

   ```xml
   <dependencies>
       <dependency>
           <groupId>org.jetbrains.kotlinx</groupId>
           <artifactId>kotlinx-serialization-json</artifactId>
           <version>${serialization.version}</version>
       </dependency>
   </dependencies>
   ```

> ### Add the Kotlin serialization library to a multiplatform project

Reviewer: We also mention Proguard rules (specific for Android) in the README.md. Do you plan to add them here?

Reviewer: I think it's a bad idea. It is not always related to custom serializer, regular `@Serializable` classes can have named companions as well. I would expect separate page for Proguard then (linked here)

To use Kotlin serialization in multiplatform projects, add the JSON serialization library dependency to your common source set:

```kotlin
commonMain {
   dependencies {
      implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:%serializationVersion%")
   }
}
```

## Serialize objects to JSON

> Serialization is the process of converting an object into a format that can be easily stored or transmitted, such as JSON.

Reviewer: This sentence repeats the one the whole document is started with

In Kotlin, you can serialize objects to JSON using the `kotlinx.serialization` library.

To make a class serializable, you need to mark it with the [`@Serializable`](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-serializable/) annotation.
This annotation instructs the compiler to generate the code required for serializing and deserializing instances of the class.
For more information, see [The @Serialization annotation](serialization-customization-options.md#the-serializable-annotation).

Let's look at an example:

1. Import the necessary serialization libraries:

    ```kotlin
    import kotlinx.serialization.*
    import kotlinx.serialization.json.*
    ```

2. Make a class serializable by annotating it with `@Serializable`:

    ```kotlin
    @Serializable
    data class Data(val a: Int, val b: String)
    ```

   > The `@Serializable` annotation enables default serialization of all properties in the primary constructor.

>    > You can customize serialization behavior with custom constructors, optional properties, and more.

Reviewer: What 'custom constructors' you mean here? I don't quite understand.

   > 
   > For more information, see [Serialize classes](serialization-customization-options.md).
   >
   {style="note"}

3. Use the [`Json.encodeToString()`](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json/encode-to-string.html) function to serialize an instance of this class:

    ```kotlin

>     // Imports the necessary libraries for serialization and JSON handling

Reviewer: Same comment about declarations here

[//]: # (title: Get started with Kotlin serialization)

> [Serialization](serialization.md) converts objects into a format that can be stored or transmitted and later reconstructed.

Reviewer 3: ```suggestion
[Serialization](serialization.md) converts objects into a format you can store or transmit and later reconstruct.
```
What do you think about making this more active?

                    </goals>
                </execution>
            </executions>
            <configuration>
                <compilerPlugins>
                    <plugin>kotlinx-serialization</plugin>
                </compilerPlugins>
            </configuration>
            <dependencies>
                <dependency>
                    <groupId>org.jetbrains.kotlin</groupId>
                    <artifactId>kotlin-maven-serialization</artifactId>
                    <version>${kotlin.version}</version>
                </dependency>
            </dependencies>
        </plugin>
    </plugins>
</build>

<dependencies>
    <dependency>
        <groupId>org.jetbrains.kotlinx</groupId>
        <artifactId>kotlinx-serialization-json</artifactId>
        <version>${serialization.version}</version>
    </dependency>
</dependencies>
```

</tab>
</tabs>

> To set up the Kotlin compiler plugin for Bazel, follow the example from the [rules_kotlin repository](https://github.com/bazelbuild/rules_kotlin/tree/master/examples/plugin/src/serialization).
> Bazel isn't officially supported by the Kotlin team, and this repository is maintained independently.
>
{style="tip"}

> ### Add the Kotlin serialization library to a multiplatform project

Reviewer 3: ```suggestion
### Add the library to a multiplatform project
```

### Configure R8 for Kotlin serialization in Android projects {initial-collapse-state="collapsed" collapsible="true"}

The Kotlin serialization library includes default [ProGuard rules](https://github.com/Kotlin/kotlinx.serialization/blob/master/rules/common.pro), so you don't need additional setup to keep serializers for all serializable classes after shrinking.
These rules don't apply to classes with named companion objects.

To retain serializers for classes with named companion objects, add rules based on the [compatibility mode](https://r8.googlesource.com/r8/+/refs/heads/master/compatibility-faq.md) you use to your `proguard-rules.pro` file:

<tabs>
<tab id="compatibility" title="R8 compatibility mode">

```bash
# Serializer for classes with named companion objects are retrieved using getDeclaredClasses
# If you have any such classes, replace the examples below with your own
-keepattributes InnerClasses # Required for getDeclaredClasses

-if @kotlinx.serialization.Serializable class
com.example.myapplication.HasNamedCompanion, # <-- List serializable classes with named companions
com.example.myapplication.HasNamedCompanion2
{
    static **$* *;
}

> -keepnames class <1>$$serializer { # Using -keepnames is enough; class is kept when serializer() is kept

Author: If I understand correctly, it means you only need to preserve the class name for the `serializer()` call, so you don’t need a stronger keep rule than -keepnames?

We could perhaps write it as : 

> Using -keepnames is enough for the serializer() call to reference the class correctly

@sandwwraith does this sound correct?

> -keepnames class <1>$$serializer { # Using -keepnames is enough; class is kept when serializer() is kept

Reviewer: LGTM

{
    static **$* *;
}
-keepnames class <1>$$serializer { # Using -keepnames is enough; class is kept when serializer() is kept
    static <1>$$serializer INSTANCE;
}

# Keep both serializer and serializable classes to save the attribute InnerClasses
-keepclasseswithmembers, allowshrinking, allowobfuscation, allowaccessmodification class
com.example.myapplication.HasNamedCompanion, # <-- List serializable classes with named companions
com.example.myapplication.HasNamedCompanion2
{
    *;
}
```

</tab>
</tabs>

> You can exclude serializable classes that are never serialized at runtime by using custom ProGuard rules with narrower [class specifications](https://www.guardsquare.com/manual/configuration/usage).
> 
{style="tip"}

## Serialize objects to JSON

In Kotlin, you can serialize objects to JSON using the `kotlinx.serialization` library.

To make a class serializable, you need to mark it with the [`@Serializable`](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-serializable/) annotation.
This annotation instructs the compiler to generate the code required for serializing and deserializing instances of the class.

> For more information, see [The @Serialization annotation](serialization-customization-options.md#the-serializable-annotation).

Reviewer: `@Serializable` btw

Let's look at an example:

1. Import declarations from the necessary serialization libraries:

    ```kotlin
    import kotlinx.serialization.*
    import kotlinx.serialization.json.*
    ```

2. Make a class serializable by annotating it with `@Serializable`:

    ```kotlin
    @Serializable

>     data class Data(val a: Int, val b: String)

Author: sure thing 👍  — rewrote both examples with those examples.

    ```

   > The `@Serializable` annotation enables default serialization of all properties with backing fields.
   > You can customize serialization behavior with property-level annotations, optional properties, and more.
   > 
   > For more information, see [Serialize classes](serialization-customization-options.md).
   >
   {style="note"}

3. Use the [`Json.encodeToString()`](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json/encode-to-string.html) function to serialize an instance of this class:

    ```kotlin
    // Imports declarations from the serialization and JSON handling libraries
    import kotlinx.serialization.*
    import kotlinx.serialization.json.*

    // Marks the Data class as serializable
    @Serializable
    data class Data(val a: Int, val b: String)

    fun main() {
        // Serializes an instance of the Data class into a JSON string
        val json = Json.encodeToString(Data(42, "str"))
        println(json)
        // {"a":42,"b":"str"}
    }
    ```

>    {kotlin-runnable="true"}

Reviewer 3: Can you give it an ID so we can see how many people run it in GA?

Author: added: id="serialize-get-started"

    ```kotlin
    import kotlinx.serialization.*
    import kotlinx.serialization.json.*
    ```

2. Make a class serializable by annotating it with `@Serializable`:

    ```kotlin
    @Serializable
    data class Data(val a: Int, val b: String)
    ```

3. Use the [`Json.decodeFromString()`](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json/decode-from-string.html) function to deserialize an object from JSON:

    ```kotlin
    // Imports declarations from the serialization and JSON handling libraries
    import kotlinx.serialization.*
    import kotlinx.serialization.json.*

    // Marks the Data class as serializable
    @Serializable
    data class Data(val a: Int, val b: String)

    fun main() {
        // Deserializes a JSON string into an instance of the Data class
        val obj = Json.decodeFromString<Data>("""{"a":42, "b": "str"}""")
        println(obj)
        // Data(a=42, b=str)
    }
    ```
   {kotlin-runnable="true"}

Congratulations! You have successfully serialized an object to JSON and deserialized it back into an object in Kotlin!

## What's next

> * Learn how to serialize standard types, including built-in types like numbers and strings, in [Serialize built-in types](serialization-serialize-builtin-types.md).

Author: hmm 🤔 yea - but I think we can remove it as a whole to make it more concise: 

> * Learn how to serialize built-in types like numbers and strings, in [Serialize built-in types](serialization-serialize-builtin-types.md).

> * Discover how to customize class serialization and adjust the default behavior of the `@Serializable` annotation in the [Serialize classes](serialization-customization-options.md) section.

Reviewer 3: Will it be a section in this page? Or is it a page itself?

Author: no the page itself so I rewrite it like: 

* Discover how to customize class serialization and adjust the default behavior of the `@Serializable` annotation in [Serialize classes](serialization-customization-options.md).
